activerecord/test/cases/validations/numericality_validation_test.rb

```diff
@@ -81,17 +81,45 @@ def self.name
     assert_predicate(subject, :valid?)
   end
 
-  def test_virtual_attribute_with_precision
+  def test_virtual_attribute_with_precision_round_down
     model_class.attribute(:virtual_decimal_number, :decimal, precision: 5)
     model_class.validates_numericality_of(
       :virtual_decimal_number, equal_to: 123.45
     )
 
-    subject = model_class.new(virtual_decimal_number: 123.455)
+    subject = model_class.new(virtual_decimal_number: 123.454)
 
     assert_predicate subject, :valid?
   end
 
+  def test_virtual_attribute_with_precision_round_half_even
+    model_class.attribute(:virtual_decimal_number, :decimal, precision: 5)
+    model_class.validates_numericality_of(
+      :virtual_decimal_number, equal_to: 123.45
+    )
+
+    subject = model_class.new(virtual_decimal_number: 123.455)
+
+    if RUBY_VERSION > "3.0.0"
```

Comment:
While this makes sense from a Ruby perspective, I think it would be better not to switch on the assertions, but on the behaviour in the validator. Less Rails apps will be broken when they upgrade ruby versions this way if we keep the same casting expectations across versions.

Reply:
I was thinking that, but I think the new `BigDecimal` code is more correct. If anything, we should backport the fix for older `BigDecimal` versions. But then that would break all Rails apps (or fix the bug for all Rails apps, depending on how you look at it).

If we do decide to keep it consistent across all ruby versions, we'll want to also update other places where we call `BigDecimal(float, precision). For example https://github.com/rails/rails/blob/master/activemodel/lib/active_model/type/decimal.rb#L46

Reply:
You're right, the old behaviour is a bug. This looks good to me. Thanks!